doc/babel.1
.Dd July 4, 2008
.Os "Open Babel" 2.2
.Dt babel 1 URM
.Sh NAME
.Nm babel , obabel
.Nd "a converter for chemistry and molecular modeling data files"
.Sh SYNOPSIS
.Nm
.Op Fl H Ar help-options
.Nm
.Op Ar OPTIONS
.Op Fl i Ar input-type
.Ar infile
.Op Fl o Ar output-type
.Ar outfile
.Pp
.Nm obabel
.Op Fl H Ar help-options
.Nm obabel
.Op Ar OPTIONS
.Op Fl i Ar input-type | Fl : Qo Ar SMILES-string Qc
.Ar infile
.Op Fl o Ar output-type
.Fl O Ar outfile
.Sh DESCRIPTION
.Nm
is a cross-platform program designed to interconvert
between many file formats used in molecular modeling and computational
chemistry and related areas.
.Pp
.Nm obabel
and
.Nm
are slightly different. The first is closer to the normal Unix convention
for commandline programs and more flexible when the user needs to specify
parameter values on options. With
.Nm babel
this only works when the option is the last on the line; with
.Nm obabel
no such restriction applies. It further has a shortcut for entering SMILES
strings, which can be used in place of an input file.
.Pp
Open Babel is also a complete programmers toolkit for developing
chemistry software. For more information, se the Open Babel web pages
\%<\fBhttp://openbabel.org/\fR>.
.Sh OPTIONS
If only input and output files are given, Open Babel will guess the
file type from the filename extension.
.Bl -tag -width flag
.It Fl : Qo Ar SMILES-string Qc
Enter SMILES string and use it in place of an input file. The SMILES string
should be enclosed in quotation marks. More than one can be used, and a molecule
title can be included if enclosed in quotes.
.It Fl a Ar options
Format-specific input options. See
.Fl H Ar format-ID
for options allowed by a particular format
.It Fl -addtotitle
Append text to the current molecule title
.It Fl -addformula
Append the molecular formula after the current molecule title
.It Fl b
Convert dative bonds: e.g., [N+]([O-])=O to N(=O)=O
.It Fl c
Center atomic coordinates at (0,0,0)
.It Fl C
Combine molecules in first file with others having the same name
.It Fl e
Continue after errors
.It Fl d
Delete Hydrogens
.It Fl --errorlevel Ar 2
Filter the level of errors and warnings displayed:
.br
       1 = critical errors only
       2 = include warnings too (default)
       3 = include informational messages too
       4 = include "audit log" messages of changes to data
       5 = include debugging messages too
.It Fl f Ar #
For multiple entry input, start import with molecule # as the first entry
.It Fl F
Output the available fingerprint types
.It Fl h
Add hydrogens
.It Fl H
Output usage information
.It Fl H Ar format-ID
Output formatting information and options for format specified
.It Fl Hall
Output formatting information and options for all formats
.It Fl i<format-ID>
Specifies input format, see below for the available formats
.It Fl j
.It Fl -join
Join all input molecules into a single output molecule entry
.It Fl k
Translate computational chemistry modeling keywords (e.g., GAMESS and Gaussian)
.It Fl m
Produce multiple output files, to allow:
.Bl -dash -offset indent -compact
.It
Splitting one input file - put each molecule into consecutively
numbered output files
.It
Batch conversion - convert each of multiple input files into a
specified output format
.El
.It Fl l Ar #
For multiple entry input, stop import with molecule # as the last entry
.It Fl o Ar format-ID
Specifies output format, see below for the available formats
.It Fl O Ar outfile
Specify the output file. This option applies to
.Nm obabel
only.
.It Fl p
Add Hydrogens appropriate for pH (use transforms in phmodel.txt)
.It Fl -property
Add or replace a property (e.g., in an MDL SD file)
.It Fl s Ar SMARTS
Convert only molecules matching the SMARTS pattern specified
.It Fl -separate
Separate disconnected fragments into individual molecular records
.It Fl t
All input files describe a single molecule
.It Fl -title Ar title
Add or replace molecular title
.It Fl x Ar options
Format-specific output options. See
.Fl H Ar format-ID
for options allowed by a particular format
.It Fl v Ar SMARTS
Convert only molecules \fBNOT\fP matching SMARTS pattern specified
.It Fl V
Output version number and exit
.It Fl z
Compress the output with gzip
.El
.Sh "FILE FORMATS"
The following formats are currently supported by Open Babel:
.Bl -item -offset indent -compact
.It
acr -- Carine ASCI Crystal
.It
alc -- Alchemy format
.It
arc -- Accelrys/MSI Biosym/Insight II CAR format [Read-only]
.It
bgf -- MSI BGF format
.It
box -- Dock 3.5 Box format
.It
bs -- Ball and Stick format
.It
c3d1 -- Chem3D Cartesian 1 format
.It
c3d2 -- Chem3D Cartesian 2 format
.It
caccrt -- Cacao Cartesian format
.It
cache -- CAChe MolStruct format [Write-only]
.It
cacint -- Cacao Internal format [Write-only]
.It
can -- Canonical SMILES format
.It
car -- Accelrys/MSI Biosym/Insight II CAR format [Read-only]
.It
ccc -- CCC format [Read-only]
.It
cdx -- ChemDraw binary format [Read-only]
.It
cdxml -- ChemDraw CDXML format
.It
cht -- Chemtool format [Write-only]
.It
cif -- Crystallographic Information File
.It
cml --  Chemical Markup Language
.It
cmlr --  CML Reaction format
.It
com -- Gaussian Cartesian Input [Write-only]
.It
copy -- Copies raw text [Write-only]
.It
crk2d -- Chemical Resource Kit 2D diagram format
.It
crk3d -- Chemical Resource Kit 3D format
.It
csr -- Accelrys/MSI Quanta CSR format [Write-only]
.It
cssr -- CSD CSSR format [Write-only]
.It
ct -- ChemDraw Connection Table format
.It
dmol -- DMol3 coordinates format
.It
ent -- Protein Data Bank format
.It
fa -- FASTA format [Write-only]
.It
fasta -- FASTA format [Write-only]
.It
fch -- Gaussian formatted checkpoint file format [Read-only]
.It
fchk -- Gaussian formatted checkpoint file format [Read-only]
.It
fck -- Gaussian formatted checkpoint file format [Read-only]
.It
feat -- Feature format
.It
fh -- Fenske-Hall Z-Matrix format [Write-only]
.It
fix -- SMILES FIX format [Write-only]
.It
fpt -- Fingerprint format [Write-only]
.It
fract -- Free Form Fractional format
.It
fs -- Open Babel FastSearching database
.It
fsa -- FASTA format [Write-only]
.It
g03 -- Gaussian 98/03 Output [Read-only]
.It
g98 -- Gaussian 98/03 Output [Read-only]
.It
gam -- GAMESS Output [Read-only]
.It
gamin -- GAMESS Input [Write-only]
.It
gamout -- GAMESS Output [Read-only]
.It
gau -- Gaussian Cartesian Input [Write-only]
.It
gjc -- Gaussian Cartesian Input [Write-only]
.It
gjf -- Gaussian Cartesian Input [Write-only]
.It
gpr -- Ghemical format
.It
gr96 -- GROMOS96 format [Write-only]
.It
gzmat -- Gaussian Z-Matrix Input
.It
hin -- HyperChem HIN format
.It
inchi -- IUPAC InChI [Write-only]
.It
inp -- GAMESS Input [Write-only]
.It
ins -- ShelX format [Read-only]
.It
jin -- Jaguar input format [Write-only]
.It
jout -- Jaguar output format [Read-only]
.It
mdl -- MDL MOL format
.It
mmd -- MacroModel format
.It
mmod -- MacroModel format
.It
mol -- MDL MOL format
.It
mol2 -- Sybyl Mol2 format
.It
molreport -- Open Babel molecule report [Write-only]
.It
moo -- MOPAC Output format [Read-only]
.It
mop -- MOPAC Cartesian format
.It
mopcrt -- MOPAC Cartesian format
.It
mopin -- MOPAC Internal
.It
mopout -- MOPAC Output format [Read-only]
.It
mpc -- MOPAC Cartesian format
.It
mpd -- Sybyl descriptor format [Write-only]
.It
mpqc -- MPQC output format [Read-only]
.It
mpqcin -- MPQC simplified input format [Write-only]
.It
nw -- NWChem input format [Write-only]
.It
nwo -- NWChem output format [Read-only]
.It
pc --  PubChem format  [Read-only]
.It
pcm --  PCModel format
.It
pdb -- Protein Data Bank format
.It
pov -- POV-Ray input format [Write-only]
.It
pqs -- Parallel Quantum Solutions format
.It
prep -- Amber Prep format [Read-only]
.It
qcin -- Q-Chem input format [Write-only]
.It
qcout -- Q-Chem output format [Read-only]
.It
report -- Open Babel report format [Write-only]
.It
res -- ShelX format [Read-only]
.It
rxn -- MDL RXN format
.It
sd -- MDL MOL format
.It
sdf -- MDL MOL format
.It
smi -- SMILES format
.It
sy2 -- Sybyl Mol2 format
.It
tdd -- Thermo format
.It
test -- Test format [Write-only]
.It
therm -- Thermo format
.It
tmol -- TurboMole Coordinate format
.It
txyz -- Tinker MM2 format [Write-only]
.It
unixyz -- UniChem XYZ format
.It
vmol -- ViewMol format
.It
xed -- XED format [Write-only]
.It
xml --  General XML format [Read-only]
.It
xyz -- XYZ cartesian coordinates format
.It
yob -- YASARA.org YOB format
.It
zin -- ZINDO input format [Write-only]
.El
.Sh "FORMAT OPTIONS"
Individual file formats may have additional formatting options.

NEWS
Open Babel 2.5.0 (2017-XX-XX)

Open Babel 2.4.0 (2016-9-15)

Open Babel 2.3.0 (2010-10-23)

This release represents a major update and should be a stable upgrade,
strongly recommended for all users of Open Babel. Highlights include a
completely rewritten stereochemistry engine, Spectrophore fingerprint
generation, 2D depiction, improved 3D coordinate generation, conformer
searching, and more. Many formats are improved or added, including
CIF, PDBQT, SVG, and more. Improved developer API and scripting
support and many, many bug fixes are also included.

What's New from 2.2.3:

* Completely rewritten stereochemistry perception, including support
  for tetrahedral, square planar, and higher-order stereochemistry.
* Dramatically improved canonicalization algorithm (Note that in general, canonical SMILES have changed since the 2.2.x release.)
* 2D depiction, including SVG vector graphics generation using code from MCDL.
* New Spectrophore generation, contributed by Silicos NV.
* New ChargeMethod API including support for partial charge assignment
  from Gasteiger, MMFF94, QEq, QTPIE methods and plugin interface for
  adding more.
* Improved 3D coordinate generation.
* New conformer generation framework, including support for diverse
  conformer generation and genetic algorithm lowest-energy searching.
* Improved user documentation.
* Improved aromaticity / Kekule bond assignment.
* Improved unit test suite using the CMake-based CTest program.
* Improved support for crystallographic unit cells (e.g., in CIF format).
* Improved UFF force field method, including hypervalent 5, 6, 7 and higher coordination numbers.
* Support for the GAFF (Generalized Amber Force Field) method.
* Support for reading geometry optimizations as multiple conformers
  from Gaussian, GAMESS-US, and other quantum chemistry packages.
* Support for reading molecular orbital energies from quantum chemistry formats.
* Several memory leaks fixed.
* Fixed many compiler warnings.
* Fixed support for MinGW and Cygwin environments.
* Fixed bugs with Gaussian 09 output files.
* Latest released version of the InChI library (1.0.3) generating standard InChI.
* Many more bug fixes and small feature improvements.

==New Command-Line Operations==

* --canonical: Output atoms in canonical order for any file format (i.e., not just SMILES)
* --conformer: Run a conformer search on the input molecules (has many options)
* --gen2D: Generate a 2D depiction of the molecule
* --partialcharge <model>: Use the partial charge model supplied to
  generate charges (i.e., instead of default Gasteiger sigma model)
* --sort <descriptor>: Sort molecules by a specified descriptor
* --unique: Only output unique molecules (as determined by InChI generation)

==New File Formats==

Import & Export:
- DL-POLY CONFIG
- FHIaims XYZ
- PDBQT

Import only:
- DL-POLY HISTORY
- GULP output
- PWscf output
- Text

Export only:
- MNA (Multilevel Neighborhoods of Atoms)
- SVG vector graphics


Open Babel 2.2.3 (2009-07-31)

What's New from 2.2.2:

This release represents an important bug-fix upgrade, strongly recommended
for all users of Open Babel.

- Fixed bug in fingerprints in 2.2.2, where the default fingerprint
  format and bit length was changed inadvertently.
- Fixed detection of shared_ptr in tr1/memory.
- Fixed additional aromaticity / Kekule assignment bugs.
- Fixed several bugs in the MMCIF format.
- Additional bug fixes.

Open Babel 2.2.2 (2009-07-04)

What's New from 2.2.1:

This release represents a major bug-fix release and is a stable
upgrade, strongly recommended for all users of Open Babel. While there
may not be many new features, many crashes and other bugs have been
fixed since 2.2.1.

- Upgraded to the new InChI 1.02 release to produce standardized InChI
  and InChIKey output.
- Fixed many stereochemistry errors when reading/writing SMILES. This
  is part of a larger project which will be finished in the 2.3
  release.
- Fixed compilation and installation on Cygwin and MinGW platforms.
- Significantly improved aromaticity and Kekule bond assignment.
- Improved 2D -> 3D coordinate generation
- Improved coordinate generation using the --gen3d command-line
  operation
- Improved performance for coordinate generation.

tools/babel.cpp
/**********************************************************************
main.cpp - Main conversion program, command-line handling.

Copyright (C) 1998-2001 by OpenEye Scientific Software, Inc.
Some portions Copyright (C) 2001-2006 by Geoffrey R. Hutchison
Some portions Copyright (C) 2004-2006 by Chris Morley

This file is part of the Open Babel project.
For more information, see <http://openbabel.org/>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation version 2 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
***********************************************************************/

// used to set import/export for Cygwin DLLs
#ifdef WIN32
#define USING_OBDLL
#endif

#include <openbabel/babelconfig.h>

#include <iostream>
#include <fstream>
#include <sstream>

#include <string>
#include <map>
#if HAVE_CONIO_H
	#include <conio.h>
#endif

#if !HAVE_STRNCASECMP
extern "C" int strncasecmp(const char *s1, const char *s2, size_t n);
#endif

#include <openbabel/obconversion.h>
#include <openbabel/plugin.h>
#include <cstdlib>
#include <cstring>

using namespace std;
using namespace OpenBabel;

void DoOption(const char* p, OBConversion& Conv, OBConversion::Option_type typ,
	      int& arg, int argc, char *argv[]);
void usage();
void help();

// There isn't a great way to do this -- we need to save argv[0] for usage()
static char *program_name;

int main(int argc,char *argv[])
{
  OBConversion Conv(&cin, &cout); //default input and output are console

  OBFormat* pInFormat = NULL;
  OBFormat* pOutFormat = NULL;
  bool inGzip = false;
  bool outGzip = false;
  vector<string> FileList, OutputFileList;
  string OutputFileName;

  // Parse commandline
  bool gotInType = false, gotOutType = false;
  bool SplitOrBatch=false;

  char *oext = NULL;
  char *iext = NULL;

  //load plugs to fully initialize option parameters
  OBPlugin::LoadAllPlugins();

  //Save name of program without its path (and .exe)
  string pn(argv[0]);
  string::size_type pos;
#ifdef _WIN32
  pos = pn.find(".exe");
  if(pos!=string::npos)
    argv[0][pos]='\0';
#endif
  pos = pn.find_last_of("/\\");
  if(pos==string::npos)
    program_name=argv[0];
  else
    program_name=argv[0]+pos+1;

  const char* p;
  int arg;
  for (arg = 1; arg < argc; ++arg)
    {
      if (argv[arg])
        {
          if (argv[arg][0] == '-')
            {
              switch (argv[arg][1])
                {

                case 'V':
                  {
                    cout << "Open Babel " << BABEL_VERSION << " -- "
                         << __DATE__ << " -- " << __TIME__ << endl;
                    exit(0);
                  }

                case 'i':
                  gotInType = true;
                  iext = argv[arg] + 2;
                  if(!*iext)
                    iext = argv[++arg]; //space left after -i: use next argument

                  if (strncasecmp(iext, "MIME", 4) == 0)
                    {
                      // get the MIME type from the next argument
                      iext = argv[++arg];
                      pInFormat = Conv.FormatFromMIME(iext);
                    }
                  else
                    {
                      //The ID provided by the OBFormat class is used as the identifying file extension
                      pInFormat = Conv.FindFormat(iext);
                    }
                  if(pInFormat==NULL)
                    {
                      cerr << program_name << ": cannot read input format!" << endl;
                      usage();
                      exit(1);
                    }
                  break;

                case 'o':
                  gotOutType = true;
                  oext = argv[arg] + 2;
                  if(!*oext)
                    oext = argv[++arg]; //space left after -i: use next argument

                  if (arg >= argc)
                    {
                      usage(); // error in parsing command-line
                      exit(1);
                    }

                  if (strncasecmp(oext, "MIME", 4) == 0)
                    {
                      // get the MIME type from the next argument
                      oext = argv[++arg];
                      pOutFormat = Conv.FormatFromMIME(oext);
                    }
                  else
                    pOutFormat = Conv.FindFormat(oext);

                  if(pOutFormat==NULL)
                    {
                      cerr << program_name << ": cannot write output format!" << endl;
                      usage();
                      exit(1);
                    }
                  break;

                case 'L': //display a list of plugin type or classes
                  {
                    const char* param=NULL;
                    if(argc>arg+1)
                      param = argv[arg+2];

                    // First assume first arg is a plugin type and
                    // param is a subtype, like babel -L ops gen3D
                    // or first arg is a plugin ID, like babel -L cml
                    OBPlugin* plugin;
                    if ((OBPlugin::GetPlugin("plugins", argv[arg+1]) &&
                         (plugin = OBPlugin::GetPlugin(argv[arg+1], param))) ||
                        (plugin = OBPlugin::GetPlugin(NULL, argv[arg+1])))
                    {
                      //Output details of subtype
                      string txt;
                      plugin->Display(txt, "verbose", argv[arg+1]);
                      cout << "One of the " << plugin->TypeID() << '\n' << txt << endl;
                      return 0;
                    }
                    //...otherwise assume it is a plugin type, like babel -L forcefields
                    //Output list of subtypes
                    OBPlugin::List(argv[arg+1], param);
                    return 0;
                  }
                case '?':
                case 'H':
                  if(isalnum(argv[arg][2]) || arg==argc-2)
                    {
                      if(strncasecmp(argv[arg]+2,"all",3))
                        {
                          const char* pID= (arg==argc-2) ? argv[arg+1] : argv[arg]+2;
                          OBFormat* pFormat = Conv.FindFormat(pID);
                          if(pFormat)
                            {
                              cout << pID << "  " << pFormat->Description() << endl;
                              if(pFormat->Flags() & NOTWRITABLE)
                                cout << " This format is Read-only" << endl;
                              if(pFormat->Flags() & NOTREADABLE)
                                cout << " This format is Write-only" << endl;

                              if(strlen(pFormat->SpecificationURL()))
                                cout << "Specification at: " << pFormat->SpecificationURL() << endl;
                            }
                          else
                            cout << "Format type: " << pID << " was not recognized" <<endl;
                        }
                      else
                        {
                          OBPlugin::List("formats","verbose");
                        }
                    }
                  else
                    help();
                  return 0;

                case '-': //long option --name text
                  {
                    //Do nothing if name is empty
                    //Option's text is the next arg provided it doesn't start with -
                    char* nam = argv[arg]+2;
                    if(*nam != '\0')
                      {
                        string txt;
                        int i;
                        for(i=0; i<Conv.GetOptionParams(nam, OBConversion::GENOPTIONS)
                              && arg<argc-1 && argv[arg+1];++i) //removed  && *argv[arg+1]!='-'
                          {
                            if(!txt.empty()) txt+=' ';
                            txt += argv[++arg];
                          }
                        if(*nam=='-')
                          {
                            // Is a API directive, e.g.---errorlevel
                            //Send to the pseudoformat "obapi" (without any leading -)
                            OBConversion apiConv;
                            OBFormat* pAPI= OBConversion::FindFormat("obapi");
                            if(pAPI)
                              {
                                apiConv.SetOutFormat(pAPI);
                                apiConv.AddOption(nam+1, OBConversion::GENOPTIONS, txt.c_str());
                                apiConv.Write(NULL, &std::cout);
                              }
                          }
                        else
                          // Is a long option name, e.g --addtotitle
                          Conv.AddOption(nam,OBConversion::GENOPTIONS,txt.c_str());
                      }
                  }
                  break;

                case 'm': //multiple output files
                  SplitOrBatch=true;
                  break;

                case 'a': //single character input option
                  p = argv[arg]+2;
                  DoOption(p,Conv,OBConversion::INOPTIONS,arg,argc,argv);
                  break;

                case 'x': //single character output option
                  p = argv[arg]+2;
                  DoOption(p,Conv,OBConversion::OUTOPTIONS,arg,argc,argv);
                  break;

                default: //single character general option
                  p = argv[arg]+1;
                  DoOption(p,Conv,OBConversion::GENOPTIONS,arg,argc,argv);
                  break;
                }
            }
          else
            {
              //filenames
              if(!gotOutType)
                FileList.push_back(argv[arg]);
              else
                OutputFileName = argv[arg];
            }
        }
    }

  if(!gotOutType) //the last file is the output
    {
      if(FileList.empty())
        {
          cerr << "No output file or format spec!" << endl;
          usage();
          exit(1);
        }
      OutputFileName = FileList.back();
      FileList.pop_back();
    }

#if defined(_WIN32) && defined(USING_DYNAMIC_LIBS)
  //Expand wildcards in input filenames and add to FileList
  vector<string> tempFileList(FileList);
  FileList.clear();
  vector<string>::iterator itr;
  for(itr=tempFileList.begin();itr!=tempFileList.end();++itr)
    DLHandler::findFiles (FileList, *itr);
#endif

  if (!gotInType)
    {
      if(FileList.empty())
        {
          cerr << "No input file or format spec!" <<endl;
          usage();
          exit(1);
        }
    }

  if (!gotOutType)
    {
      pOutFormat = Conv.FormatFromExt(OutputFileName.c_str(), outGzip);
      if(pOutFormat==NULL)
        {
          cerr << program_name << ": cannot write output format!" << endl;
          usage();
          exit(1);
        }
    }

    if(!Conv.SetInFormat(pInFormat))
    {
      cerr << "Invalid input format" << endl;
      usage();
      exit(1);
    }
    if(!Conv.SetOutFormat(pOutFormat, outGzip))
    {
      cerr << "Invalid output format" << endl;
      usage();
      exit(1);
    }

  if(SplitOrBatch)
    {
      //Put * into output file name before extension (or ext.gz)
      if(OutputFileName.empty())
        {
          OutputFileName = "*.";
          if (oext != NULL)
            OutputFileName += oext;
        }
      else
        {
          string::size_type pos = OutputFileName.rfind(".gz");
          if(pos==string::npos)
            pos = OutputFileName.rfind('.');
          else
            pos = OutputFileName.rfind('.',pos-1);
          if(pos==string::npos)
            OutputFileName += '*';
          else
            OutputFileName.insert(pos,"*");
        }
    }

  int count = Conv.FullConvert(FileList, OutputFileName, OutputFileList);

  Conv.ReportNumberConverted(count);

  if(OutputFileList.size()>1)
    {
      clog << OutputFileList.size() << " files output. The first is " << OutputFileList[0] <<endl;
    }

  std::string messageSummary = obErrorLog.GetMessageSummary();
  if (messageSummary.size())
    {
      clog << messageSummary << endl;
    }

#ifdef _DEBUG
  //CM keep window open
  cout << "Press any key to finish" <<endl;
  getch();
#endif

  return 0;
}

void DoOption(const char* p, OBConversion& Conv,
	      OBConversion::Option_type typ, int& arg, int argc, char *argv[])
{
  while(p && *p) //can have multiple single char options
  {
    char ch[2]="?";
    *ch = *p++;
    const char* txt=NULL;
    //Get the option text if needed
    int nParams = Conv.GetOptionParams(ch, typ);
    if(nParams)
    {
      if(*p)
      {
        txt = p; //use text immediately following the option letter
        p=NULL; //no more single char options
      }
      else if(arg<argc-1)
      {
        txt = argv[++arg]; //use text from next arg
        if(*txt=='-')
        {
          //...unless it is another option
          cerr << "Option -" << ch << " takes a parameter" << endl;
          exit(0);
        }
      }
    }
    Conv.AddOption(ch, typ, txt);
  }
}

void deprecation()
{
  cout << "WARNING: babel is deprecated and will be removed in a future release" << endl
    << "         of Open Babel. You should use obabel instead. For information" << endl
    << "         on the differences please see:"
    << endl << "             http://openbabel.org/docs/current/Command-line_tools/babel.html"
    << endl;
}

void usage()
{
  cout << "Open Babel " << BABEL_VERSION << " -- " << __DATE__ << " -- "
       << __TIME__ << endl;
  cout << "Usage: " << program_name
       << " [-i<input-type>] <name> [-o<output-type>] <name>" << endl;
  cout << "Try  -H option for more information." << endl;
  cout << endl;
  deprecation();

#ifdef _DEBUG
  //CM keep window open
  cout << "Press any key to finish" <<endl;
  getch();
#endif
}

void help()
{
  cout << "Open Babel converts chemical structures from one file format to another"<< endl << endl;
  cout << "Usage: " << program_name << " <input spec> <output spec> [Options]" << endl << endl;

  deprecation();
  cout << endl;

  cout << "Each spec can be a file whose extension decides the format." << endl;
  cout << "Optionally the format can be specified by preceding the file by" << endl;
  cout << "-i<format-type> e.g. -icml, for input and -o<format-type> for output" << endl << endl;
  cout << "See below for available format-types, which are the same as the " << endl;
  cout << "file extensions and are case independent." << endl;
  cout << "If no input or output file is given stdin or stdout are used instead." << endl << endl;
  cout << "More than one input file can be specified and their names can contain" <<endl;
  cout << "wildcard chars (* and ?).The molecules are aggregated in the output file.\n" << endl;
  cout << OBConversion::Description(); // Conversion options
  cout << "-H Outputs this help text" << endl;
  cout << "-V Outputs version number" <<endl;
  cout << "-L <category> Lists plugin classes of this category, e.g. <formats>" << endl;
  cout << "   Use just -L for a list of plugin categories." << endl;
  cout << "   Use -L <ID> e.g. -L sdf for details of a format or other plugin." << endl;
  cout << "-m Produces multiple output files, to allow:" <<endl;
  cout << "   Splitting: e.g.        " << program_name << " infile.mol new.smi -m" <<endl;
  cout << "     puts each molecule into new1.smi new2.smi etc" <<endl;
  cout << "   Batch conversion: e.g. " << program_name << " *.mol -osmi -m" <<endl;
  cout << "     converts each input file to a .smi file" << endl;
#ifdef _WIN32
  cout << "   In Windows these can also be done using the forms" <<endl;
  cout << "     " << program_name << " infile.mol new*.smi and " << program_name << " *.mol *.smi respectively.\n" <<endl;
#endif

  OBFormat* pDefault = OBConversion::GetDefaultFormat();
  if(pDefault)
    cout << pDefault->TargetClassDescription();// some more options probably for OBMol

  OBFormat* pAPI= OBConversion::FindFormat("obapi");
  if(pAPI)
    cout << pAPI->Description();

  cout << "To see a list of recognized file formats use\n  babel -L formats\n"
       << "To see details and specific options for a particular format, e.g CML, use\n  babel -L cml\n"
       << endl;
  //cout << "The following file formats are recognized:" << endl;
  //OBPlugin::List("formats");
  //cout << "\nSee further specific info and options using -H<format-type>, e.g. -Hcml" << endl;
}

/* OpenBabel man page*/
/** \page babel a converter for chemistry and molecular modeling data files
*
* \n
* \par SYNOPSIS
*
* \b babel [-H<help-options>] [-V] [-m] [-d] [-h] [-p] [-s<SMARTS-pattern>] [-v<SMARTS-pattern>] [-f<#> -l<#>] [-c] [-x<format-options>] [-i<input-type>] \<infile\> [-o<output-type>] \<outfile\>
*
* \par DESCRIPTION
*
* Open Babel is a program designed to interconvert a number of
* file formats currently used in molecular modeling software. \n\n
*
* Note that Open Babel can also be used as a library to interconvert
* many file formats and to provide standard chemistry software routines.
* See the Open Babel web pages (http://openbabel.org) for more
* information.
*
* \par OPTIONS
*
* If only input and output files are given, Open Babel will guess
* the file type from the filename extension. \n\n
*
* \b -V :
*     Output version number and exit \n\n
* \b -H :
*     Output usage information \n\n
* \b -H\<format-ID\> :
*     Output formatting information and options for format specified\n\n
* \b -Hall :
*     Output formatting information and options for all formats\n\n
* \b -i :
*     Specifies input format, see below for the available formats \n\n
* \b -o :
*     Specifies output format, see below for the available formats \n\n
* \b -m :
*     Produce multiple output files, to allow:\n
*      * Splitting one input file - put each molecule into consecutively numbered output files \n
*      * Batch conversion - convert each of multiple input files into a specified output format \n
*     See examples below \n\n
* \b -d :
*     Delete Hydrogens \n\n
* \b -h :
*     Add Hydrogens \n\n
* \b -p :
*     Add Hydrogens appropriate for pH (use transforms in phmodel.txt)  \n\n
* \b -t :
*     All input files describe a single molecule \n\n
* \b -f\<#\> :
*     For multiple entries input, start import at molecule # \n\n
* \b -l\<#\> :
*     For multiple entries input, stop import at molecule # \n\n
* \b -c :
*     Center atomic coordinates at (0,0,0) \n\n
* \b -s\<SMARTS\> :
*     Convert only molecules matching the SMARTS pattern specified \n\n
* \b -v\<SMARTS\> :
*     Convert only molecules \b NOT matching SMARTS pattern specified \n\n
*
* \par FILE FORMATS
*
* The following formats are currently supported by Open Babel:
*  \n    alc -- Alchemy format
*  \n    bgf -- BGF format
*  \n    box -- Dock 3.5 Box format
*  \n    bs -- Ball and Stick format
*  \n    c3d1 -- Chem3D Cartesian 1 format
*  \n    c3d2 -- Chem3D Cartesian2 format
*  \n    caccrt -- Cacao format
*  \n    cache -- CAChe format [Writeonly]
*  \n    cacint -- CacaoInternal format [Writeonly]
*  \n    car -- MSI Biosym/Insight II CAR format [Readonly]
*  \n    ccc -- CCC format [Readonly]
*  \n    cht -- ChemTool format [Writeonly]
*  \n    cml -- Chemical Markup Language
*  \n    com -- Gaussian Cartesian Input [Writeonly]
*  \n    crk2d -- Chemical Resource Kit diagram format (2D)
*  \n    crk3d -- Chemical Resource Kit 3D format
*  \n    csr -- CSR format [Writeonly]
*  \n    cssr -- CSD CSSR format [Writeonly]
*  \n    ct -- ChemDraw Connection Table format
*  \n    dmol -- DMol3 coordinates format
*  \n    ent -- Protein Data Bank format
*  \n    feat -- Feature format
*  \n    fh -- Fenske-Hall Z-Matrix format [Writeonly]
*  \n    fix -- FIX format [Writeonly]
*  \n    g03 -- Gaussian 98/03 Output [Readonly]
*  \n    g98 -- Gaussian 98/03 Output [Readonly]
*  \n    gam -- GAMESS Output
*  \n    gamout -- GAMESS Output
*  \n    gau -- Gaussian Cartesian Input [Writeonly]
*  \n    gjc -- Gaussian Cartesian Input [Writeonly]
*  \n    gjf -- Gaussian Cartesian Input [Writeonly]
*  \n    gpr -- Ghemical format
*  \n    gr96 -- GROMOS96 format [Writeonly]
*  \n    gzmat -- Gaussian Z-Matrix Input
*  \n    hin -- HyperChem Input format
*  \n    ins -- ShelX format [Readonly]
*  \n    jout -- Jaguar output format
*  \n    mdl -- MDL MOL format
*  \n    mm1gp -- Ghemical format
*  \n    mm3 -- MM3 format [Writeonly]
*  \n    mmd -- MacroMod format
*  \n    mmod -- MacroMod format
*  \n    mol -- MDL MOL format
*  \n    mol2 -- Sybyl Mol2 format
*  \n    mopcrt -- MOPAC Cartesian format
*  \n    mopout -- MOPAC Output format [Readonly]
*  \n    mpqc -- MPQC format [Readonly]
*  \n    nwo -- NWChem format
*  \n    pdb -- Protein Data Bank format
*  \n    pov -- POV-Ray input format [Writeonly]
*  \n    pqs -- Parallel Quantum Solutions format
*  \n    prep -- Amber Prep format [Readonly]
*  \n    qcout -- QChem output format
*  \n    qm1gp -- Ghemical format
*  \n    report -- Report format [Writeonly]
*  \n    res -- ShelX format [Readonly]
*  \n    rxn -- MDL RXN format
*  \n    sd -- MDL MOL format
*  \n    sdf -- MDL MOL format
*  \n    smi -- SMILES format
*  \n    tmol -- TurboMole Coordinate format
*  \n    txyz -- Tinker format [Writeonly]
*  \n    unixyz -- UniChem XYZ format
*  \n    vmol -- ViewMol format
*  \n    xed -- XED format [Writeonly]
*  \n    xyz -- XYZ cartesian coordinates format
*  \n    zin -- ZINDO input format [Writeonly]
*
* \par FORMAT OPTIONS
*  Individual file formats may have additional formatting options. \n
*  Input format options are preceded by 'a', e.g. -as \n
*  Output format options are preceded by 'x', e.g. -xn \n
*    For further specific information and options, use -H<format-type> \n
*    e.g., -Hcml
*
* \par EXAMPLES
*  - Standard conversion \n
*     babel -ixyz ethanol.xyz -opdb ethanol.pdb \n
*  - Conversion from a SMI file in STDIN to a Mol2 file written to STDOUT \n
*     babel -ismi -omol2 \n
*  - Split a multi-molecule file into new1.smi, new2.smi, etc. \n
*     babel infile.mol new.smi -m \n
*
* \par AUTHORS
*
* Open Babel is currently maintained by \b Geoff \b Hutchison, \b Chris \b Morley and \b Michael \b Banck.
*
* For more contributors to Open Babel, see http://openbabel.org/THANKS.shtml
*
* \par COPYRIGHT
*  Copyright (C) 1998-2001 by OpenEye Scientific Software, Inc.
*  Some portions Copyright (C) 2001-2005 by Geoffrey R. Hutchison \n \n
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation version 2 of the License.\n \n
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
* \par SEE ALSO
*   The web pages for Open Babel can be found at http://openbabel.org/
**/
